### PR TITLE
Eurostat education attainment

### PR DESCRIPTION
**Pull request by eftekhari-mhs**

### CODE REVIEW

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/Eurostats_NUTS2_Edat.mcf [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+measuredProperty: dcs:count
+statType: dcs:measuredValue
+measurementDenominator: dcs:Count_Person_25To64Years
+
+# Population aged 25-64 by educational attainment level (3-4)
+Node: dcid:Count_Person_25To64Years_UpperSecondaryEducationOrPostSecondaryNonTertiaryEducation_AsAFractionOfCount_Person_25To64Years
+typeOf: dcs:StatisticalVariable
+populationType: dcs:Person
+age: [Years 25 64]
+educationalAttainment: dcs:UpperSecondaryEducation&PostSecondaryNonTertiaryEducation
```
Comment (jeffreyoldham) [inline scripts/eurostat/regional_statistics_by_nuts/education_attainment/Eurostats_NUTS2_Edat.mcf:26]: Is the "&" correct?

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/Eurostats_NUTS2_Edat.mcf [hunk 2]
```diff
@@ -0,0 +1,129 @@
+# Population aged 25-64 by educational attainment level (0-2)
+Node: dcid:Count_Person_25To64Years_LessThanPrimaryEducationOrPrimaryEducationOrLowerSecondaryEducation_AsAFractionOfCount_Person_25To64Years
+typeOf: dcs:StatisticalVariable
+populationType: dcs:Person
+age: [Years 25 64]
+educationalAttainment: dcs:LessThanPrimaryEducation&PrimaryEducation&LowerSecondaryEducation
```
Comment (jeffreyoldham) [inline scripts/eurostat/regional_statistics_by_nuts/education_attainment/Eurostats_NUTS2_Edat.mcf:6]: Is the "&" correct?

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/README.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+1. [About the Dataset](#about-the-dataset)
+1. [Import Artifacts](#import-artifacts)
+1. [Generating Artifacts](#generating-artifacts)
+
+## About the Dataset
+
+### Download URL
+
+[TSV] file is available for download from [here](https://ec.europa.eu/eurostat/estat-navtree-portlet-prod/BulkDownloadListing?file=data/edat_lfse_04.tsv.gz).
```
Comment (jeffreyoldham) [inline scripts/eurostat/regional_statistics_by_nuts/education_attainment/README.md:14]: Similar to #266.

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/README.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+2. sex,edu-level,age,unit,geo: e.g. F,ED0-2,Y25-64,PC,AT (Female, Lower Secondary or less,Years 25 to 64, Percent, nuts/AT)
+3. value: some with note "e"="estimated", "u" = "low reliability", and "d"="definition differs"
+
+sex,edu-level,age,unit,geo is further broken down into:
+1. sex-level: 'F_ED0-2', 'F_ED3-8', 'F_ED3_4','F_ED5-8', 'M_ED0-2', 'M_ED3-8', 'M_ED3_4', 'M_ED5-8', 'T_ED0-2', 'T_ED3-8', 'T_ED3_4', 'T_ED5-8'
+2. age: 25 to 64 years old
+3. unit: Percentage
+4. geo: NUTS2 codes for regions of Europe
+
+The cleraned csv is broken up into 5 major families of variables:
```
Comment (jeffreyoldham) [inline scripts/eurostat/regional_statistics_by_nuts/education_attainment/README.md:46]:
```suggestion
The cleaned csv is broken up into 5 major families of variables:

```

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/README.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+14. Count_Person_25To64Years_TertiaryEducation_Male_AsAFractionOfCount_Person_25To64Years_Male: float values 
+
+
+### Notes and Caveats
+
+- There are breaks in the time series of the original dataset. Note =”b”, value=":"
+- There are estimated values in the original dataset. Note =”e”
+- There are values with low reliability. Note = "u"
+- There are values that definition differs. Note = "d"
+- As written, we're imporing the estimated, low reliable, and different definitions values as regular values.
```
Comment (jeffreyoldham) [inline scripts/eurostat/regional_statistics_by_nuts/education_attainment/README.md:69]:
```suggestion
- As written, we're importing the estimated, low reliable, and different definitions values as regular values.

```

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/README.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+
+### License
+
+Eurostat has a policy of encouraging free re-use of its data, both for non-commercial and commercial purposes. 
+
+The license is available online at [here](https://ec.europa.eu/eurostat/about/policies/copyright).
```
Comment (jeffreyoldham) [inline scripts/eurostat/regional_statistics_by_nuts/education_attainment/README.md:76]: Similar to #266.

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/README.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+#### StatisticalVariable Instance MCF
+[Eurostats_NUTS2_Edat.mcf](./Eurostats_NUTS2_Edat.mcf)
+
+#### Scripts
+[education_attainment_preprocess_gen_tmcf.py](./education_attainment_preprocess_gen_tmcf.py)
+
+#### Pre-Processing Validation
+
+Automated validation:
+1. In the processing script (next section), there assertion to check numerical values are correctly formatted.
```
Comment (jeffreyoldham) [inline scripts/eurostat/regional_statistics_by_nuts/education_attainment/README.md:104]: Similar to #266.

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/csv_template_mcf_compatibility_checker.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pandas as pd
+
+
+def test_col_names(cleaned_csv, tmcf):
+    """ check if all the column names specified in the template mcf
```
Comment (jeffreyoldham) [inline scripts/eurostat/regional_statistics_by_nuts/education_attainment/csv_template_mcf_compatibility_checker.py:19]: Similar to #266.

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/config.txt [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  import_name: "Eurostats_NUTS2_Education_Attainment_by_Gender"
+  provenance_url: "https://datcom-browser-dev2.googleplex.com/kg?dcid=dc/kzsldr"
+  provenance_description: "The education attainment level for EU countries down to NUTS2 regions"
+  # Path to the resolved MCF. Can be a repeated field.
+  mcf_url: "https://storage.cloud.google.com/resolved_mcf/external_tables/Eurostats_NUTS2_Education_Attainment_by_Gender/nodes.mcf-00000-of-00001"
+  enums: "https://storage.cloud.google.com/import_server/dev/Enum_nodes_Education_attainment/Enum_nodes_Education_attainment.mcf"
+  # First date of imported data.
+  start_date: "2020-01-01"
+  # Last date of imported data.
+  end_date: "2020-14-07"
```
Comment (jeffreyoldham) [inline scripts/eurostat/regional_statistics_by_nuts/education_attainment/config.txt:11]: Similar to #266.

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/config.txt [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  # Source data release schedule.
+  update_frequency: "P1Y"
+  next_release_date: "2021-05-14"
+}
+© 2020 G
```
Comment (jeffreyoldham) [inline scripts/eurostat/regional_statistics_by_nuts/education_attainment/config.txt:16]:
```suggestion
```

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/Eurostats_NUTS2_Edat.mcf (hunk 1 above)
Comment (eftekhari-mhs) [reply scripts/eurostat/regional_statistics_by_nuts/education_attainment/Eurostats_NUTS2_Edat.mcf:26]: @jeffreyoldham Yes, this is how they decided to aggregate different enums
import doc: https://docs.google.com/document/d/192NoZd_-O5SfsHGkBiHteWbgjqCZkD8LUcSfAE_Icak/edit#

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/Eurostats_NUTS2_Edat.mcf (hunk 2 above)
Comment (eftekhari-mhs) [reply scripts/eurostat/regional_statistics_by_nuts/education_attainment/Eurostats_NUTS2_Edat.mcf:6]: same :)

**Review by jeffreyoldham — COMMENTED** [review]
Thanks.  Please see the suggested changes in #266.  We are close to finishing.  Thanks.

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/education_attainment_preprocess_gen_tmcf.py [hunk 11]
```diff
@@ -16,12 +16,9 @@
...
 
-# _DATA_URL = "https://ec.europa.eu/eurostat/estat-navtree-portlet-prod/BulkDownloadListing?file=data/edat_lfse_04.tsv.gz"
-
+_DATA_URL = "https://ec.europa.eu/eurostat/estat-navtree-portlet-prod/BulkDownloadListing?file=data/edat_lfse_04.tsv.gz"
 _CLEANED_CSV = "./Eurostats_NUTS2_Edat.csv"
 _TMCF = "./Eurostats_NUTS2_Edat.tmcf"
-_SOURCE_TSV = "./edat_lfse_04.tsv"
-_SOURCE_CSV_LONG = "./edat_lfse_04.csv"
 
 _OUTPUT_CLOUMNS = [
```
Comment (jeffreyoldham) [inline scripts/eurostat/regional_statistics_by_nuts/education_attainment/education_attainment_preprocess_gen_tmcf.py:23]:
```suggestion
_OUTPUT_COLUMNS = [
```

File: scripts/eurostat/regional_statistics_by_nuts/education_attainment/education_attainment_preprocess_gen_tmcf.py [hunk 12]
```diff
@@ -80,68 +77,49 @@ def translate_wide_to_long(source_tsv, source_csv_long):
...
-    df.to_csv(source_csv_long, index=False)
+    return df
 
 
-def preprocess(cleaned_csv, source_csv_long, output_columns):
+def preprocess(df, cleaned_csv):
     with open(cleaned_csv, 'w', newline='') as f_out:
         writer = csv.DictWriter(f_out,
-                                fieldnames=output_columns,
+                                fieldnames=_OUTPUT_CLOUMNS,
```
Comment (jeffreyoldham) [inline scripts/eurostat/regional_statistics_by_nuts/education_attainment/education_attainment_preprocess_gen_tmcf.py:86]:
```suggestion
                                fieldnames=_OUTPUT_COLUMNS,
```

**Review by jeffreyoldham — APPROVED** [review]
Please correct the spelling of "columns" and then we will submit.  Thanks.

**Review by jeffreyoldham — APPROVED** [review]
Thanks.

**Pull request merged**